app/org/sagebionetworks/bridge/dynamodb/DynamoUserConsentDao.java

```diff
@@ -6,81 +6,127 @@
-import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapperConfig.SaveBehavior;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBMapper.FailedBatch;
+import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBQueryExpression;
 import com.amazonaws.services.dynamodbv2.datamodeling.DynamoDBScanExpression;
 import com.amazonaws.services.dynamodbv2.model.AttributeValue;
 import com.amazonaws.services.dynamodbv2.model.ComparisonOperator;
 import com.amazonaws.services.dynamodbv2.model.Condition;
-import com.amazonaws.services.dynamodbv2.model.ConditionalCheckFailedException;
 import com.google.common.collect.Sets;
 
+/**
+ * Currently migrating to a new consent table. The options for keeping the two tables in sync:
+ * 
+ *  table3 succeeds, table2 succeeds - fine
+ *  table3 fails,    --              - fine, return error
+ *  table3 succeeds, table2 fails    - we can fix this in migration, or undo the table3 record.
+ *                                     Also, return an error
+ */
 @Component
 public class DynamoUserConsentDao implements UserConsentDao {
 
-    private DynamoDBMapper mapper;
+    private DynamoDBMapper mapperV2;
+    private DynamoDBMapper mapperV3;
+
+    @Resource(name = "userConsentDdbMapper2")
+    public final void setDdbMapper2(DynamoDBMapper mapper2) {
+        this.mapperV2 = mapper2;
+    }
 
-    @Autowired
-    public void setDynamoDbClient(BridgeConfig bridgeConfig, AmazonDynamoDB client) {
-        DynamoDBMapperConfig mapperConfig = new DynamoDBMapperConfig.Builder().withSaveBehavior(SaveBehavior.UPDATE)
-                .withConsistentReads(ConsistentReads.CONSISTENT)
-                .withTableNameOverride(DynamoUtils.getTableNameOverride(DynamoUserConsent2.class, bridgeConfig)).build();
-        mapper = new DynamoDBMapper(client, mapperConfig);
+    @Resource(name = "userConsentDdbMapper3")
+    public final void setDdbMapper3(DynamoDBMapper mapper3) {
+        this.mapperV3 = mapper3;
     }
 
     @Override
     public UserConsent giveConsent(String healthCode, StudyConsent studyConsent) {
         checkArgument(isNotBlank(healthCode), "Health code is blank or null");
         checkNotNull(studyConsent);
-        DynamoUserConsent2 consent = null;
-        try {
-            consent = getUserConsent(healthCode, studyConsent);
-            if (consent == null) {
-                consent = new DynamoUserConsent2(healthCode, studyConsent);
+
+        // This will be passed in to the method when we coordinate Stormpath and DDB operations
+        long signedOn = DateUtils.getCurrentMillisFromEpoch();
+
+        // The logic is being changed here. You cannot update the signing date on an active
+        // consent record if it exists. Throw an exception if it does.
+        DynamoUserConsent2 consent2 = getUserConsent2(healthCode, studyConsent.getStudyKey());
+        if (consent2 != null) {
+            throw new EntityAlreadyExistsException(null, "Consent already exists.");
+        }
```

Comment:
Can we test this end-to-end through the HTTP API? I suspect this will actually through a NullPointerException in EntityAlreadyExistsException.getEntityClass() when Play tries to serialize the exception, but don't know for certain. Either way, a null check in getEntityClass() would be a good idea.